channels/tsmf/client/tsmf_codec.c

```diff
@@ -411,6 +524,114 @@ static UINT32 tsmf_codec_parse_VIDEOINFOHEADER(TS_AM_MEDIA_TYPE* mediatype, wStr
 	return 48;
 }
 
+
+const static int avpriv_mpeg4audio_sample_rates[16] = {
+    96000, 88200, 64000, 48000, 44100, 32000,
+    24000, 22050, 16000, 12000, 11025, 8000, 7350
+};
+
+static int get_sample_rates(int idx){
+	if(idx<0 || idx>sizeof(avpriv_mpeg4audio_sample_rates)/sizeof(avpriv_mpeg4audio_sample_rates[0])){
+		WLog_ERR(TAG, "get_sample_rates: invalid index:%d", idx);
+		return -1;
+	}
+	return avpriv_mpeg4audio_sample_rates[idx];
+}
+
+static UINT32 tsmf_codec_parse_HEAACWAVEINFOHEADER(TS_AM_MEDIA_TYPE* mediatype, wStream* s)
+{
+	/*
+		typedef struct heaacwaveformat_tag {
+		  HEAACWAVEINFO wfInfo;
+		  BYTE          pbAudioSpecificConfig[1];
+		} HEAACWAVEFORMAT, *PHEAACWAVEFORMAT;
+		
+		typedef struct heaacwaveinfo_tag {
+		  WAVEFORMATEX wfx;
+		  WORD         wPayloadType;
+		  WORD         wAudioProfileLevelIndication;
+		  WORD         wStructType;
+		  WORD         wReserved1;
+		  DWORD        dwReserved2;
+		} HEAACWAVEINFO, *PHEAACWAVEINFO;
+	*/
+	UINT16 wPayloadType;
+	UINT16 wAudioProfileLevelIndication;
+	UINT16 wStructType;
+
+	Stream_Read_UINT16(s, wPayloadType);
```

Comment:
You should not Stream_Read() without checking there's enough to read (Stream_GetRemainingLength), so I guess your function should return a BOOL to say if the decoding was ok, and the actual return value should be passed as a pointer on a UINT32. The remark on Stream_Read applies for the rest of the function.